aptos-move/framework/aptos-framework/sources/staking_contract.move

```diff
@@ -0,0 +1,1132 @@
+        update_distribution_pool(
+            distribution_pool, total_distribution_amount, operator, staking_contract.commission_percentage);
+
+        pool_u64::buy_in(distribution_pool, recipient, coins_amount);
+        let pool_address = staking_contract.pool_address;
+        emit_event(
+            add_distribution_events,
+            AddDistributionEvent { operator, pool_address, amount: coins_amount }
+        );
+    }
+
+    // Calculate accumulated rewards and commissions since last update.
+    fun get_staking_contract_amounts_internal(staking_contract: &StakingContract): (u64, u64, u64) {
+        // Pending_inactive is not included in the calculation because pending_inactive can only come from:
+        // 1. Outgoing commissions. This means commission has already been extracted.
+        // 2. Stake withdrawals from stakers. This also means commission has already been extracted as
+        // request_commission_internal is called in unlock_stake
+        let (active, _, pending_active, _) = stake::get_stake(staking_contract.pool_address);
+        let total_active_stake = active + pending_active;
+        let accumulated_rewards = total_active_stake - staking_contract.principal;
+        let commission_amount = accumulated_rewards * staking_contract.commission_percentage / 100;
+
+        (total_active_stake, accumulated_rewards, commission_amount)
+    }
+
+    fun create_stake_pool(
+        staker: &signer,
+        operator: address,
+        voter: address,
+        contract_creation_seed: vector<u8>,
+    ): (signer, SignerCapability, OwnerCapability)  {
+        // Generate a seed that will be used to create the resource account that hosts the staking contract.
+        let seed = bcs::to_bytes(&signer::address_of(staker));
+        vector::append(&mut seed, bcs::to_bytes(&operator));
+        // Include a salt to avoid conflicts with any other modules out there that might also generate
+        // deterministic resource accounts for the same staker + operator addresses.
+        vector::append(&mut seed, SALT);
+        // Add an extra salt given by the staker in case an account with the same address has already been created.
+        vector::append(&mut seed, contract_creation_seed);
+
+        let (stake_pool_signer, stake_pool_signer_cap) = account::create_resource_account(staker, seed);
+        stake::initialize_stake_owner(&stake_pool_signer, 0, operator, voter);
+
+        // Extract owner_cap from the StakePool, so we have control over it in the staking_contracts flow.
+        // This is stored as part of the staking_contract. Thus, the staker would not have direct control over it without
+        // going through well-defined functions in this module.
+        let owner_cap = stake::extract_owner_cap(&stake_pool_signer);
+
+        (stake_pool_signer, stake_pool_signer_cap, owner_cap)
+    }
+
+    fun update_distribution_pool(
```

Comment:
this function is to re-distribute shares in case the pending_inactive grows? it's very confusing to me, I've been staring at it for over 20 mins

Reply:
Yes it's a bit complex because every other shareholder in the distribution pool has to pay the operator the commission on rewards earned on pending_inactive stake (can stick around for the rest of the recurring lockup period).

The key part is this line https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/staking_contract.move#L640 which computes the unpaid commission based on the current value of the distribution pool (with accumulated rewards) which converts that to shares that the other shareholders have to pay the operator (they can only pay in shares).